modules/frontend/pipeline/responses.go

```diff
@@ -87,10 +94,27 @@ func (a *asyncResponse) SendError(err error) {
 	}
 }
 
-func (a *asyncResponse) done() {
+// SendComplete indicates the sender is done. We close the channel to give a clear signal to the consumer
+func (a *asyncResponse) SendComplete() {
 	close(a.respChan)
 }
 
+// NextComplete indicates the receiver is done. We drain all channels and subchannels to goroutines are orphaned
+func (a *asyncResponse) NextComplete() {
+	go func() {
+		// drain the response channel?
+		for resps := range a.respChan {
```

Comment:
Can you check my understanding: The goroutine leak was that if a receiver only partially read the response channel (because it hit max search results for example), then a sender would be orphaned waiting on the channel send.  Now when the receiver calls NextComplete it drains the channel until.

Reply:
correct, per an out of band conversation i have moved towards using the cancel context signal from the `Next()`er to cancel the Sends.

This is a bit more hidden and easy to break, but it's also cleaner and more in line with channel architectures.

Reply:
👍 lgtm thanks